server/src/main/scala/za/co/absa/atum/server/model/CheckpointItemWithPartitioningFromDB.scala

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2021 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.atum.server.model
+
+import io.circe.{DecodingFailure, Json}
+import za.co.absa.atum.model.dto.{
+  CheckpointWithPartitioningDTO,
+  MeasureDTO,
+  MeasureResultDTO,
+  MeasurementDTO,
+  PartitionDTO,
+  PartitioningWithIdDTO
+}
+
+import java.time.ZonedDateTime
+import java.util.UUID
+
+case class CheckpointItemWithPartitioningFromDB(
+  idCheckpoint: UUID,
+  checkpointName: String,
+  author: String,
+  measuredByAtumAgent: Boolean,
+  measureName: String,
+  measuredColumns: Seq[String],
+  measurementValue: Json, // JSON representation of `MeasurementDTO`
+  checkpointStartTime: ZonedDateTime,
+  checkpointEndTime: Option[ZonedDateTime],
+  idPartitioning: Long,
+  partitioning: Json,
+  partitioningAuthor: String,
+  hasMore: Boolean
+)
+
+object CheckpointItemWithPartitioningFromDB {
+
+  private def fromItemsToCheckpointWithPartitioningDTO(
+    checkpointItems: Seq[CheckpointItemWithPartitioningFromDB]
+  ): Either[Throwable, CheckpointWithPartitioningDTO] = {
+    val measurementsOrErr = checkpointItems.map { checkpointItem =>
```

Comment:
Hmm, this function feels like harder to track.
Wouldn't be possible to write it as a for comprehension with more subroutines doing each part (more private functions).
The other option seemed to me, to maybe use a fold on the collection (again shortened by private functions).
🤔

Reply:
Also `collectFirst` would come handy.